erpnext/selling/doctype/quotation/quotation.py

```diff
@@ -38,6 +39,14 @@ def validate_valid_till(self):
 		if self.valid_till and getdate(self.valid_till) < getdate(self.transaction_date):
 			frappe.throw(_("Valid till date cannot be before transaction date"))
 
+	def validate_shopping_cart_items(self):
+		if self.order_type != "Shopping Cart":
+			return
+
+		for item in self.items:
+			if not frappe.db.exists("Website Item", {"item_code": item.item_code}):
+				frappe.throw(_("Item {0} must be a website item for Shopping Cart quotations").format(item.item_code))
```

Comment:
```suggestion
            has_web_item = frappe.db.exists("Website Item", {"item_code": item.item_code})
            template = frappe.db.get_value("Item", item.item_code, "variant_of")
            if not has_web_item and template:
                has_web_item = frappe.db.exists("Website Item", {"item_code": has_template})
                
			if not has_web_item:
				frappe.throw(_("Row #{0}: Item {1} must be a Website Item for Shopping Cart quotations").format(item.idx, item.item_code))
```

**The condition here needs to _not_ throw the validation if it is a variant item whose template is published.**

Item variants without website items are allowed to be added to cart. I've encountered that some users do not create individual web pages for variants at all, and here Website Item seems wasteful.
Reference: https://github.com/frappe/erpnext/pull/29383

Fixing this should fix the failing server side test on this PR

Reply:
This is good to know. I never use variants so I didn't even think of this.